kubeadm upgrade plan: implement structured output

bart0sh opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This is a continuation of the series of PRs implementing structured output in kubeadm described in this [KEP](https://github.com/kubernetes/enhancements/pull/1054)

The PR implements structured output for ```kubeadm upgrade plan``` command in JSON and YAML formats.

Example output:

- Default text output - the same as in the current kubeadm version 

```
# kubeadm upgrade plan
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:[upgrade/versions] Cluster version: v1.16.0
[upgrade/versions] kubeadm version: v1.16.0-alpha.0.6718+082262b9f8e4c1-dirty
[upgrade/versions] Latest stable version: v1.16.3
[upgrade/versions] Latest version in the v1.16 series: v1.16.3

Components that must be upgraded manually after you have upgraded the control plane with 'kubeadm upgrade apply':

COMPONENT   CURRENT       AVAILABLE
kubelet     1 x v1.16.0   v1.16.3

Upgrade to the latest version in the v1.16 series:

COMPONENT                 CURRENT   AVAILABLE
kube-apiserver            v1.16.0   v1.16.3
kube-controller-manager   v1.16.0   v1.16.3
kube-scheduler            v1.16.0   v1.16.3
kube-proxy                v1.16.0   v1.16.3
coredns                   1.6.2     1.6.2
etcd                      3.3.17    3.3.17-0

You can now apply the upgrade by executing the following command:

	kubeadm upgrade apply v1.16.3

Note: Before you can perform this upgrade, you have to update kubeadm to v1.16.3.

_____________________________________________________________________
```

- JSON

```
# kubeadm upgrade plan -o json
{
    "kind": "UpgradePlan",
    "apiVersion": "output.kubeadm.k8s.io/v1alpha1",
    "components": [
        {
            "name": "kubelet",
            "currentVersion": "1 x v1.16.0",
            "newVersion": "v1.16.3"
        },
        {
            "name": "kube-apiserver",
            "currentVersion": "v1.16.0",
            "newVersion": "v1.16.3"
        },
        {
            "name": "kube-controller-manager",
            "currentVersion": "v1.16.0",
            "newVersion": "v1.16.3"
        },
        {
            "name": "kube-scheduler",
            "currentVersion": "v1.16.0",
            "newVersion": "v1.16.3"
        },
        {
            "name": "kube-proxy",
            "currentVersion": "v1.16.0",
            "newVersion": "v1.16.3"
        },
        {
            "name": "coredns",
            "currentVersion": "1.6.2",
            "newVersion": "1.6.2"
        },
        {
            "name": "etcd",
            "currentVersion": "3.3.17",
            "newVersion": "3.3.17-0"
        }
    ]
}

```
- YAML

```
# kubeadm upgrade plan -o yaml
apiVersion: output.kubeadm.k8s.io/v1alpha1
components:
- currentVersion: 1 x v1.16.0
  name: kubelet
  newVersion: v1.16.3
- currentVersion: v1.16.0
  name: kube-apiserver
  newVersion: v1.16.3
- currentVersion: v1.16.0
  name: kube-controller-manager
  newVersion: v1.16.3
- currentVersion: v1.16.0
  name: kube-scheduler
  newVersion: v1.16.3
- currentVersion: v1.16.0
  name: kube-proxy
  newVersion: v1.16.3
- currentVersion: 1.6.2
  name: coredns
  newVersion: 1.6.2
- currentVersion: 3.3.17
  name: etcd
  newVersion: 3.3.17-0
kind: UpgradePlan
```

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: implemented structured output of 'kubeadm upgrade plan' in JSON and YAML formats
```

bart0sh committed:
add UpgradePlan to the kubeadm.output API group

bart0sh committed:
kubeadm: extend Printer interface

- added new methods: Fprintln, Flush, Close
- added new constants JSONOutput and YAMLOutput

bart0sh committed:
kubeadm upgrade plan: implement structured output

Used cli-runtime API to print upgrade plan in 3 formats:
    - TEXT (identical to the current output)
    - YAML
    - JSON

bart0sh committed:
Test upgrade plan structured output

Implemented tests for upgrade plan output in JSON and YAML formats.

bart0sh commented:
/cc @neolit123 @fabriziopandini @rosti @akutz @kad @Klaven

yastij commented:
/cc

bart0sh commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

rosti commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +59,78 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+	}
+
+	// Print item
+	item := obj.(*outputapiv1alpha1.UpgradePlanItem)
```
It's best to check if the cast was good. Something spurious may surprise us here and panic will ensue.

bart0sh replied:
makes sense, thank you. will do.

rosti commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -92,85 +156,63 @@ func runPlan(flags *planFlags, userVersion string) error {
 	if len(upgrades) == 0 {
-		fmt.Fprintln(w, "Awesome, you're up-to-date! Enjoy!")
-		return
+		printer.Fprintf(writer, "Awesome, you're up-to-date! Enjoy!\n")
```
We should probably add `printer.Fprintln`.

bart0sh replied:
ok, will do

rosti commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -92,85 +156,63 @@ func runPlan(flags *planFlags, userVersion string) error {
+		printer.Fprintf(writer, "\nUpgrade to the latest %s:\n\n", upgrade.Description)
+
+		items := []*outputapiv1alpha1.UpgradePlanItem{
+			newUpgradePlanItem("API Server", upgrade.Before.KubeVersion, upgrade.After.KubeVersion),
```
This is interesting. Ideally, in the machine output one would like a more machine friendly component name here (`constants.KubeAPIServer`). However, that's not ideal for end users (text output), where we would want to keep `API Server`.
We can try to have both somehow and, depending on the output type, dump the preferred one?

bart0sh replied:
Yes, we can, but it would look quite ugly. We'll have to wrap all similar places into something like 
```
if out  == output.TextOutput { 
    name = "API Servier"
} else {
    name = constants.KubeaAPIServer
} 
``` everywhere we want to use this.

Any better idea how to do it?

rosti replied:
I'll think of some Go magic to hide this. You are right, that ```if...else``` everywhere is barely readable and ugly.

neolit123 replied:
using kube-apiserver for both is fine and i've seen that in other places in k/k.

bart0sh replied:
@neolit123 thank you for the review. this should be done now.

rosti reviewed (commented):
Thanks @bart0sh !

fabriziopandini commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -63,7 +61,7 @@ func (g *KubeVersionGetter) ClusterVersion() (string, *versionutil.Version, erro
 		return "", nil, errors.Wrap(err, "Couldn't fetch cluster version from the API Server")
 	}
-	fmt.Fprintf(g.w, "[upgrade/versions] Cluster version: %s\n", clusterVersionInfo.String())
+	g.printer.Printf("[upgrade/versions] Cluster version: %s\n", clusterVersionInfo.String())
```
[1] Ideally, we should get rid of the print inside the getter, and let the caller decide if to print or not the returned value. When this change merges, then the printer parameter can be removed from this func 

But I'm not going to block on this (it can be addressed on follow up PRs as well.

bart0sh replied:
Yes, I was thinking about removing this kind of things. However, considering that this function was called from different places I thought it's dangerous to remove the output as in can cause user confusion/ make current UI less informative. Moving it to all places where this function is called doesn't look as a good solution to me either.

I'd prefer to leave it as it is and address this in a separate PRs.

fabriziopandini replied:
Ok, if you open an issue to keep track of this might be we can get some help too (this can make a good first issue)

bart0sh replied:
Created an issue: https://github.com/kubernetes/kubeadm/issues/1853

fabriziopandini commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -75,7 +73,7 @@ func (g *KubeVersionGetter) ClusterVersion() (string, *versionutil.Version, erro
 func (g *KubeVersionGetter) KubeadmVersion() (string, *versionutil.Version, error) {
 	kubeadmVersionInfo := version.Get()
-	fmt.Fprintf(g.w, "[upgrade/versions] kubeadm version: %s\n", kubeadmVersionInfo.String())
+	g.printer.Printf("[upgrade/versions] kubeadm version: %s\n", kubeadmVersionInfo.String())
```
[1]

fabriziopandini commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -37,12 +35,13 @@ import (
-	fmt.Fprintf(w, "[%s] Reading configuration from the cluster...\n", logPrefix)
-	fmt.Fprintf(w, "[%s] FYI: You can look at this config file with 'kubectl -n %s get cm %s -oyaml'\n", logPrefix, metav1.NamespaceSystem, constants.KubeadmConfigConfigMap)
+func FetchInitConfigurationFromCluster(client clientset.Interface, printer output.Printer, logPrefix string, newControlPlane bool) (*kubeadmapi.InitConfiguration, error) {
+	printer.Printf("[%s] Reading configuration from the cluster...\n", logPrefix)
```
Same here [1]

fabriziopandini commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -92,85 +156,63 @@ func runPlan(flags *planFlags, userVersion string) error {
 	// Compute which upgrade possibilities there are
 	klog.V(1).Infoln("[upgrade/plan] computing upgrade possibilities")
-	availUpgrades, err := upgrade.GetAvailableUpgrades(versionGetter, flags.allowExperimentalUpgrades, flags.allowRCUpgrades, etcdClient, cfg.DNS.Type, client)
+	availUpgrades, err := upgrade.GetAvailableUpgrades(versionGetter, flags.allowExperimentalUpgrades, flags.allowRCUpgrades, etcdClient, cfg.DNS.Type, client, printer)
```
[2] It is a little bit weird that `GetAvailableUpgrades` require in input a printer, when few lines below we are calling `printAvailableUpgrades`

bart0sh replied:
Thank you for pointing this out. Will do.

fabriziopandini commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -74,8 +75,8 @@ type ClusterState struct {
 // kinds of upgrades can be performed
-func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesAllowed, rcUpgradesAllowed bool, etcdClient etcdutil.ClusterInterrogator, dnsType kubeadmapi.DNSAddOnType, client clientset.Interface) ([]Upgrade, error) {
-	fmt.Println("[upgrade] Fetching available versions to upgrade to")
+func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesAllowed, rcUpgradesAllowed bool, etcdClient etcdutil.ClusterInterrogator, dnsType kubeadmapi.DNSAddOnType, client clientset.Interface, printer output.Printer) ([]Upgrade, error) {
```
[2] I think we can remove printer here and move the two printf out of this call

bart0sh replied:
will do

fabriziopandini commented on cmd/kubeadm/app/phases/upgrade/health.go:
```diff
@@ -59,8 +59,8 @@ func (c *healthCheck) Name() string {
 // - (if static pod-hosted) that all required Static Pod manifests exist on disk
-func CheckClusterHealth(client clientset.Interface, ignoreChecksErrors sets.String) error {
-	fmt.Println("[upgrade] Making sure the cluster is healthy:")
+func CheckClusterHealth(client clientset.Interface, ignoreChecksErrors sets.String, printer output.Printer) error {
```
Same as [2] is it possible to move the printf outside this func?

fabriziopandini reviewed (changes requested):
@bart0sh Looks great!
Only few minors comment mostly oriented at keeping under control when/where the printer object flows down the call chain.

Having Printf statements everywhere is one of the technical debt existing in kubeadm, and this PR is a good chance to start tacking some on them :wink:

bart0sh commented on cmd/kubeadm/app/apis/output/types.go:
```diff
@@ -32,3 +32,15 @@ type BootstrapToken struct {
+// UpgradePlanItem item represents information about upgrade plan for one component
+type UpgradePlanItem struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
```
I think we do. At least without it I'm getting this error when buiding kubeadm:
```
cmd/kubeadm/app/apis/output/zz_generated.deepcopy.go:74:3: cannot use c (type *UpgradePlanItem) as type runtime.Object in return argument:
	*UpgradePlanItem does not implement runtime.Object (missing GetObjectKind method)
```

rosti replied:
That's probably a bad build issue. Sometimes doing an incremental build is going to cause weird errors like this.
`runtime.Object` should be implemented when the deep copy generator is invoked and the object has `metav1.TypeMeta` in it. `metav1.ObjectMeta` should have nothing to do with it.

bart0sh replied:
you're right. it was some weird build failure. I'll remove metav1.ObjectMeta from the struct. Thank you for the suggestion!

bart0sh commented on cmd/kubeadm/app/apis/output/types.go:
```diff
@@ -32,3 +32,15 @@ type BootstrapToken struct {
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradePlanItem item represents information about upgrade plan for one component
+type UpgradePlanItem struct {
```
i didn't do it for two reasons:
- it adds one level of complexity without adding any new information.
- it makes it impossible to use unbuffered output

rosti replied:
I don't think, that the lack of unbuffered output for upgrade plan is a big deal. The problem is, that we end up with a bunch of different YAML documents in a single output and that can be inconvenient to parse.

bart0sh replied:
I'd say they're pretty straightforward to parse. Here is an example: https://play.golang.com/p/BFCyyx8lx4R

Thank you for this. Playing with parsing helped me to understand that we'll probably need to add `yaml:"<name>"` tags to the public structures.

neolit123 replied:
for me, buffering makes sense only when there is considerable network delay between fetching the items. if "plan" does not have delay between printing the same items currently (i don't think it does), we can block until we have all items and have them better structured as @rosti suggests here:
https://github.com/kubernetes/kubernetes/pull/83941/files#r336393656

another problem with buffering here - what about if we want to have global fields for the "plan" output in the next version (v1alpha2) and not only a list of items? this will demand as separate struct for those...where the items are buffered and at the end we get these global fields in a separate strcture.

so, i think i'm leaning towards a +1 for https://github.com/kubernetes/kubernetes/pull/83941/files#r336393656
even if we sacrifice buffering.

yet, we are probably going to have the same discussion for other commands?

bart0sh replied:
I'll look what I can do here. It'd complicates the code it seems, but hopefully doable. I hope it's ok to make text output buffered too just for the sake of less code/consistency?

neolit123 replied:
i seem to be talking without recollecting how "upgrade plan" actually works.

```
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade] Fetching available versions to upgrade to
[upgrade/versions] Cluster version: v1.14.0
[upgrade/versions] kubeadm version: v1.14.0
...
table goes here
```

these are steps that require actions over a network...

so a problem here is the following, and i believe i spoke about this at some point.
we don't have good separation of structured output and random log entries like the above.

klog always prints to stderr, but kubeadm's sparse println calls print to stdout.
if "text" is used as the output format we don't care, but if a structured output is used, to me it makes sense the following:

- switch to print all log lines to stderr
- print the output the user cares about to stdout

this provides:
- the users can filter stdout and only care about it.
- based on exit status they can figure out if the command failed
- look at logs from stderr
- make the log output unbufferred and buffer only the structured output?

these aspects are missing from the kep...
@bart0sh we can discuss this in the kubeadm office hours if you want.

neolit123 replied:
that said, this is obviously a big change and we don't want to do it....
> switch to print all log lines to stderr

so a question here is do we want to go with the "let's buffer everything".
the problem with that is that the user will get no indication if a network call blocks for longer periods of time...

bart0sh replied:
So far I'm taking care of this step by step. For every output that I switch to structured/text I'm trying to make sure the output is not messed up with the prints. @fabriziopandini pointed out in [his comment above](https://github.com/kubernetes/kubernetes/pull/83941#pullrequestreview-304464424) that Print calls in the API should be considered as a technical debt and removing them is another good thing that's triggered by structured output implementation.

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd

bart0sh commented:
@fabriziopandini @rosti thank you for the review! I've updated the PR according to your suggestions.

Klaven reviewed (approved):
/lgtm
It's a big PR. I pulled and tested it a little bit and it worked fine. I think we should merge it and iterate on it at this point unless someone sees something breaking.

bart0sh commented:
@neolit123 @rosti @fabriziopandini 1.17 deadline is approaching. Kindly ping for review&merge.

bart0sh commented:
/test pull-kubernetes-e2e-gce

fabriziopandini commented:
I'm +1 for merging before code freeze

bart0sh commented:
@fabriziopandini I'm working on unbuffered version that @rosti and @neolit123 suggested. However, I'm not against merging this one of course. Hopefully I'll have time to update it before the release.

neolit123 commented:
> @fabriziopandini I'm working on unbuffered version that @rosti and @neolit123 suggested. However, I'm not against merging this one of course. Hopefully I'll have time to update it before the release.

i can see if i can test the PR locally. really out of time myself this week.

neolit123 commented:
@bart0sh we spoke about this during the one shot meeting today (about pending PRs, etc).
TL;DR is that we though that the buffered vs unbuffered cases don't matter much and can be fixed later.
however the API that @rosti suggested here is preferred.
https://github.com/kubernetes/kubernetes/pull/83941#discussion_r336393656

we want to do that so that we can leave the potential to expose global plan fields as part of the output structure and not limit our self only to the "item" API objects.

if you are not able to update this in time, it's fine because the backend and the images command was added, so we can still claim some alpha support.

would be still nice if we can also get this PR in with the refactor.
thanks.

bart0sh committed:
kubeadm: fix build failure

Used kubeadm Printer object in the
FetchInitConfigurationFromCluster call to fix this build issue:

   app/cmd/alpha/certs.go:271:68: cannot use out (type io.Writer)
        as type output.Printer in argument to config.FetchInitConfigurationFromCluster:
            io.Writer does not implement output.Printer (missing Close method)

bart0sh commented:
@neolit123 @rosti @fabriziopandini Updated the PR according to your comments. Please review again.

I managed to implement only YAML and JSON output so far. Both are fully buffered. Text output stays unbuffered. Please, look at the PR description to see how the output looks like.

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd

neolit123 commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -272,12 +272,18 @@ const (
 	// HyperKube defines variable used internally when referring to the hyperkube image
 	HyperKube = "hyperkube"
+	// CoreDNS defines variable used internally when referring to the CoreDNS component
+	CoreDNS = "coredns"
```
"CoreDNS"

neolit123 commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -272,12 +272,18 @@ const (
+	// CoreDNS defines variable used internally when referring to the CoreDNS component
+	CoreDNS = "coredns"
+	// KubeDNS defines variable used internally when referring to the KubeDNS component
+	KubeDNS = "kubedns"
```
"kube-dns"

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -196,32 +346,45 @@ func printAvailableUpgrades(upgrades []upgrade.Upgrade, w io.Writer, isExternalE
+		for _, item := range items {
+			err = printer.PrintObj(item, writer)
+			if err != nil {
+				return errors.Wrapf(err, "can't print upgrade plan item %v", item)
```
"Cannot print..."

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -92,85 +262,65 @@ func runPlan(flags *planFlags, userVersion string) error {
+				item := newComponentUpgradePlan(name, fmt.Sprintf("%d x %s", nodeCount, oldVersion), upgrade.After.KubeVersion)
+				err = printer.PrintObj(item, writer)
+				if err != nil {
+					return errors.Wrapf(err, "can't print upgrade plan item %v", item)
```
"Cannot print..."

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+func (upjyp *upgradePlanJSONYAMLPrinter) PrintObj(obj runtime.Object, writer io.Writer) error {
+	item, ok := obj.(*outputapiv1alpha1.ComponentUpgradePlan)
+	if !ok {
+		return fmt.Errorf("Expected ComponentUpgradePlan, but got %+v", obj)
```
Expected -> expected

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+}
+
+// AllowedFormats returns list of allowed output formats
+func (ppf *upgradePlanPrintFlags) AllowedFormats() []string {
```
this can be just `(pf *upgradePlanPrintFlags)`

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+}
+
+// AllowedFormats returns list of allowed output formats
+func (upjypf *upgradePlanJSONYamlPrintFlags) AllowedFormats() []string {
```
this can be just `(pf *upgradePlanJSONYamlPrintFlags)`

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+
+	item, ok := obj.(*outputapiv1alpha1.ComponentUpgradePlan)
+	if !ok {
+		return fmt.Errorf("Expected ComponentUpgradePlan, but got %+v", obj)
```
Expected -> expected

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+	// JSONYamlPrintFlags provides default flags necessary for json/yaml printing.
+	JSONYamlPrintFlags *upgradePlanJSONYamlPrintFlags
+	// upgradePlanTextPrintFlags provides default flags necessary for text printing.
+	TextPrintFlags *upgradePlanTextPrintFlags
```
comment above should be `// TextPrintFlag ...`

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+// upgradePlanPrintFlags defines printer flag structure for the
+// upgrade plan kubeadm command and provides a method
+// of retrieving a known printer based on flag values provided.
+type upgradePlanPrintFlags struct {
```
not a blocking comment, but 
it feels to me PrintFlags should have been an interface that different outputs implement.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/output/output.go#L41

- e.g. upgradePlanPrintFlags being an implementation of PrintFlags.

re-defining a similar structure for the different outputs doesn't seem right.

bart0sh replied:
I'm not sure I understand how defining an interface helps here. It's a local structure used only to define couple of methods. Even if I group those methods into the interface how I suppose to use that?

neolit123 replied:
it is an OOP problem that can be solved.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/output/output.go#L38-L40
with this comment we claim that this structure will be reused, but it's not used in this file.

bart0sh replied:
Yes, it will be reused at least in 'config images list' output. It can't be reused in the upgrade plan ouput due to buffered type of output that you've requested.

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -92,85 +262,65 @@ func runPlan(flags *planFlags, userVersion string) error {
-			} else {
-				UnstableVersionFlag = " --allow-experimental-upgrades"
-			}
-		}
```
moving the above doesn't seem necessary.
this can help the reviewers.

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan_test.go:
```diff
@@ -493,20 +513,22 @@ _____________________________________________________________________
+kube-controller-manager   v1.10.2   v1.11.0
+kube-scheduler            v1.10.2   v1.11.0
+kube-proxy                v1.10.2   v1.11.0
+coredns                             1.0.6
```
only requested to change to CoreDNS and kube-dns as these are official names.
the lowercase change for kube-* components seem fine.

a potential fix for us later is making this test output less volatile and not binding to constant names.

neolit123 commented:
@bart0sh i had a quick pass, no major concerns aside from rename requests.

the only thing that i think needs refactor is https://github.com/kubernetes/kubernetes/pull/83941#discussion_r346292265

in terms of the state of the feature i already requested from the release team that we can graduate "structured output" as ALPHA without this PR.

so it's not needed, but we can still merge it today.
feels a bit of risky to merge it in the last minute as we broke the plan command a couple of times in the last cycles.

in any case this needs more review from others too :eyes:

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -30,6 +30,12 @@ import (
+// JSONOutput describes the plain text output
+const JSONOutput = "json"
+
+// YAMLOutput describes the plain text output
```
the comments above YAMLOutput and JSONOuput mention "plain text"

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+	OutputFormat string
+}
+
+func newupgradePlanPrintFlags(outputFormat string) *upgradePlanPrintFlags {
```
newupgradePlan... -> newUpgradePlan

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -52,21 +61,182 @@ func NewCmdPlan(apf *applyPlanFlags) *cobra.Command {
+}
+
+// newupgradePlanJSONYAMLPrinter creates new upgradePlanJSONYAMLPrinter object
+func newupgradePlanJSONYAMLPrinter(resourcePrinter printers.ResourcePrinter, err error) (output.Printer, error) {
```
newUpgrade...

bart0sh commented:
@neolit123 thank you for the review. Updated PR.

neolit123 commented:
> Yes, it will be reused at least in 'config images list' output. It can't be reused in the upgrade plan ouput due to buffered type of output that you've requested.

i don't like how this ended up, but will defer to others for more opinions.
if we refactor this in master later, backporting that to older branches is not possible.
which makes backporting fixes difficult in the process.

bart0sh commented:
> backporting that to older branches is not possible.

Why do we need to backport internal changes if we keep the output unchanged?

neolit123 commented:
> Why do we need to backport internal changes if we keep the output unchanged?

not bugs related to the output, generic bugs in plan get buried into the refactors that will happen in master.

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd

rosti commented:
/remove-lifecycle rotten
/lifecycle frozen

neolit123 commented:
/remove-lifecycle frozen
